contracts/contracts/harvest/Harvester.sol

```diff
@@ -0,0 +1,376 @@
+            .uniswapV2CompatibleAddr;
+        rewardTokenConfigs[_tokenAddress] = tokenConfig;
+
+        IERC20 token = IERC20(_tokenAddress);
+
+        // if changing token swap provider cancel existing allowance
+        if (
+            oldUniswapAddress != address(0) &&
+            oldUniswapAddress != _uniswapV2CompatibleAddr
+        ) {
+            token.safeApprove(oldUniswapAddress, 0);
+        }
+
+        // Give Uniswap infinite approval when needed
+        if (_uniswapV2CompatibleAddr != address(0)) {
+            if (
+                token.allowance(address(this), _uniswapV2CompatibleAddr) <
+                type(uint256).max / 2
+            ) {
+                token.safeApprove(_uniswapV2CompatibleAddr, 0);
+                token.safeApprove(_uniswapV2CompatibleAddr, type(uint256).max);
+            }
+        }
+
+        emit RewardTokenConfigUpdated(
+            _tokenAddress,
+            _allowedSlippageBps,
+            _harvestRewardBps,
+            _uniswapV2CompatibleAddr,
+            _liquidationLimit
+        );
+    }
+
+    /***************************************
+                    Rewards
+    ****************************************/
+
+    /**
+     * @dev Transfer token to governor. Intended for recovering tokens stuck in
+     *      contract, i.e. mistaken sends.
+     * @param _asset Address for the asset
+     * @param _amount Amount of the asset to transfer
+     */
+    function transferToken(address _asset, uint256 _amount)
+        external
+        onlyGovernor
+    {
+        IERC20(_asset).safeTransfer(governor(), _amount);
+    }
+
+    /**
+     * @dev Collect reward tokens from all strategies
+     */
+    function harvest() external onlyGovernor nonReentrant {
+        _harvest();
+    }
+
+    /**
+     * @dev Swap all supported swap tokens for stablecoins via Uniswap.
+     */
+    function swap() external onlyGovernor nonReentrant {
+        _swap();
+    }
+
+    /*
+     * @dev Collect reward tokens from all strategies and swap for supported
+     *      stablecoin via Uniswap
+     */
+    function harvestAndSwap() external nonReentrant {
+        // TODO: add protection so that harvestAndSwap isn't called twice too closely together
+        _harvest();
+        _swap();
+        if (msg.sender != vaultAddress) {
+            IVault(vaultAddress).rebase();
+        }
+    }
+
+    /**
+     * @dev Collect reward tokens from all strategies
+     */
+    function _harvest() internal {
+        address[] memory allStrategies = IVault(vaultAddress)
+            .getAllStrategies();
+        for (uint256 i = 0; i < allStrategies.length; i++) {
+            _harvest(allStrategies[i]);
+        }
+    }
+
+    /**
+     * @dev Collect reward tokens for a specific strategy. Called from the vault.
+     * @param _strategyAddr Address of the strategy to collect rewards from
+     */
+    function harvest(address _strategyAddr) external onlyGovernor nonReentrant {
```

Comment:
We need to make sure that we do only allow the _strategyAddr to be one of our valid strategies.

We do the correctly check internally when doing _swap(address), which protects swap(address), and harvestAndSwap(address), but we also need to check it here on harvest(address). Check all my assumptions here. :)

In general we don't want to allow our contract to make calls to arbitrary external addresses unless we absolutely have to do so.

Reply:
Felt a bit of cold sweat down mu back, not thinking about strategy address potentially being malicious contract. Will add tests for this

Reply:
Added the check to `_harvest(address)` function that is called by all the public & governorOnly functions. Will also add tests for this one.

Reply:
We don't need to check that it's an active strategy on the vault, just that it's a strategy that has been configured locally and is setup on the harvester. This is cleaner and cheaper, as well as meaning that we can harvest even if the strategy is removed from the vault.

Reply:
Currently the harvester doesn't hold any info regarding the strategy addresses. How do you propose this could be achieved? Maybe Harvester having a mapping something along the lines of: 

```
// mapping of all (active or non active) strategies that ever existed
mapping(address => bool) public strategyRegistry;
```
That mapping could be set by the Vault when `approveStrategy` is called - and not removed / unset when `removeStrategy` is called. 

Or you had something else in mind?

Reply:
Yes, exactly this.

Reply:
Ran a quick test calling `harvestAndSwap(cvxAddress)` with the existing call to Vault to fetch all strategies and check if it is a valid one and without the Vault call. The gas costs were 569637 & 562516  (7.1k saving without the check). Things affecting decision of having strategies configured locally in Harvester: 
- 🟢 7.1k saved in gas costs when calling `harvestAndSwap(address)`
- 🟢 harvester is able to harvest from strategies that were once and are no longer active on the Vault
- 🔴 Vault needs to be aware of Harvester again to be able to update its state on `approveStrategy` calls 

I would rather keep a cleaner code separation and have Vault not being Harvester aware. Open to being persuaded to the contrary.

Reply:
That's a great gas savings.

I agree that it is awesome for the vault not to know about the harvester. I think it's okay if the vault doesn't configure the harvester, since that is an operation that can be done externally. 

In general my thinking is that contracts enforce properties and logic that is security critical, but outside that as much thinking that can live outside the chain lives out there. This makes the contracts have much less code in them, thus be simpler, and thus both easier to be correct, secure, and less needing of change. 

The question here is really if the vault and the harvester being out of sync is a security concern.

I don't think it is one if the new strategy is being added. We can just harvest later, nothing is running that should not be. Same non-problem if a strategy is in the vault and removed from harvesting.

The question in my mind is around risks if we remove a strategy from the vault and forget to remove a strategy from the harvester.

The harvester couldn't cause the vault a problem in such a case - it can only send the vault funds. The harvester doesn't call any state updating actions on the vault.

If a strategy has been removed from the vault, the strategy should be empty of principal funds (though it may have reward tokens laying around). We only currently allow one strategy at a time to be harvested, which means that calling harvest on one strategy should not interfere with the harvest on another strategy.

I lean towards being fine with the possible disconnect between the two. This is more a gut feeling, value call. I really like things that don't know about other things, less code, and big gas savings in our core function. I know that smart people might disagree with me. @tomlinton Got any thoughts here?

Reply:
I will let you make the call here. My primary instinct was to have better separation of concerns between different contracts, then again I have no objections on the points you are making above. Agree that one way or another there shouldn't be any negative security implications even in cases of disconnects between vault & harvester. 

Will add a mapping of strategyAddress -> bool indicating if a strategy is a valid one, that will be updated by the vault when governor approves the strategy on the vault. Will also let that property be updatable by the governor just in case we would ever need that control. But I agree with your points, that currently I don't see why that would be required.

Reply:
done via: https://github.com/OriginProtocol/origin-dollar/pull/869/commits/69f4803c88303edf0073c5789328b1d77db5e188